Added Codemaster Queue

xoxfaby opened the pull request:
Button to toggle added to panel
Extra panel added for queue
Players can leave/join queue via buttons
Top 2 in queue are put as codemaster on next start, otherwise random players.
spectators and black color removed from queue on game start

Ryan6578 commented on src/Global.-1.ttslua:
```diff
@@ -821,6 +915,90 @@ function toggleShuffle(o, color)
+end
+
+function onPlayerChangeColor(color)
+  if color == "Black" then
```
Why not iterate through spectators when a player moves to Grey and remove any names in the codemasterQueue that are now spectators?

xoxfaby replied:
Spectators and Black are already removed at the beginning of a shuffle. Otherwise I would have to keep track of whether or not we are currently in the "Player standing up phase" and ignore color changes during? Unless Player.changeColor doesn't trigger onPlayerChangeColor but I would assume it does.

Ryan6578 commented on src/Global.-1.ttslua:
```diff
@@ -821,6 +915,90 @@ function toggleShuffle(o, color)
+
+function onPlayerChangeColor(color)
+  if color == "Black" then
+    for k,v in ipairs(codemasterQueue) do
```
Assign k/v more useful names, please.

Ryan6578 commented on src/Global.-1.ttslua:
```diff
@@ -821,6 +915,90 @@ function toggleShuffle(o, color)
+end
+
+function onPlayerDisconnect(player)
+  for k,v in ipairs(codemasterQueue) do
```
Assign k/v more useful names, please.

Ryan6578 commented on src/Global.-1.ttslua:
```diff
@@ -821,6 +915,90 @@ function toggleShuffle(o, color)
+      return
+    end
+  end
+  table.insert(codemasterQueue,{Player[color].steam_id, Player[color].steam_name})
```
Just make sure this is refactored throughout the code where it appears.

xoxfaby commented on src/Global.-1.ttslua:
```diff
@@ -821,6 +915,90 @@ function toggleShuffle(o, color)
+function redrawQueue()
+  local nButtons = #queuePanel.getButtons() - 3
+  if nButtons > 0 then
+    for i = nButtons+2,3,-1 do
```
Maybe I implemented it wrong but it would run the for loop at least once no matter what. 
Also have to go from back to front because otherwise the deleting of buttons messes with the indexes of the other buttons.

xoxfaby commented on src/Global.-1.ttslua:
```diff
@@ -751,6 +828,8 @@ function startShuffle()
       "Brown"
     }
     local randSide = math.random(0,1)
+    local randMast = math.random(0,1)
```
Oh yeah, that was used for exactly that, but then I realized that since all players are shuffled, it doesn't matter which color they are, since they won't be with the same team again. 

This would only be useful with an extra mode where teams aren't shuffled.

xoxfaby commented on src/Global.-1.ttslua:
```diff
@@ -732,13 +802,20 @@ function startShuffle()
     if p.color != "Grey" and p.color != "Black" then
       table.insert(players, p)
+    else
+      for k,v in ipairs(codemasterQueue) do
```
Sure thing, not used to writing code that other people might read, will do.

xoxfaby commented:
I addressed all your requested changes. k, v has been mostly renamed to i(ndex) , player following what I'm seeing in the code already, old variables for unnecessary codemaster shuffling have been removed. 

added a variable to keep track of when we are shuffling players so that we can ignore onPlayerChangeColor during that time and started removing players from the queue if they go Grey or Black

Also fixed a few random issues and added a few comments.

Ryan6578 commented on src/Global.-1.ttslua:
```diff
@@ -733,11 +805,22 @@ function startShuffle()
       table.insert(players, p)
+    else
+      -- remove gray or black players from the codemasterQueue
+      for i, q in ipairs(codemasterQueue) do
```
Please rename i/q here.

Ryan6578 replied:
You can also rename p to just players, if you so please.

Ryan6578 commented on src/Global.-1.ttslua:
```diff
@@ -776,25 +862,40 @@ function startShuffle()
         coroutine.yield(0)
     end
 
+    shufflingPlayers = false
```
Move this to before you redraw the queue. This is so that if someone switches to Gray/Black during the 250-tick delay, it is properly reflected in the redraw. Otherwise, during this period, they will not be removed from the queue if they exist in it.

Ryan6578 commented on src/Global.-1.ttslua:
```diff
@@ -980,6 +983,9 @@ function joinQueue(o, color)
 end
 
 function leaveQueue(o, color)
+  if color == "Grey" or color == "Black" then
```
You shouldn't need to check for Grey, as Grey players can't interact with buttons either way.

Ryan6578 replied:
Also, you should output an error message. Something like:

```
Player[color].broadcast("Please switch colors before attempting to join the codemaster queue.", redColor)
```

Ryan6578 commented:
Or maybe both. But the latter just as a safeguard.

Ryan6578 commented:
Or better yet. Don't remove them from the queue, but when they sit to spectator, have the button with their name grayed out. Keep it like that for 5 minutes. After a 5 minute timer, remove their name from the queue completely @xoxfaby.

Either way, good ideas.
